token-lending/program/src/state/obligation.rs

```diff
@@ -1,186 +1,223 @@
-    pub deposited_collateral_tokens: u64,
-    /// Reserve which collateral tokens were deposited into
-    pub collateral_reserve: Pubkey,
-    /// Borrow rate used for calculating interest.
-    pub cumulative_borrow_rate_wads: Decimal,
-    /// Amount of tokens borrowed for this obligation plus interest
-    pub borrowed_liquidity_wads: Decimal,
-    /// Reserve which tokens were borrowed from
-    pub borrow_reserve: Pubkey,
-    /// Mint address of the tokens for this obligation
-    pub token_mint: Pubkey,
+    /// Last update to collateral, liquidity, or their market values
+    pub last_update: LastUpdate,
+    /// Lending market address
+    pub lending_market: Pubkey,
+    /// Owner authority which can borrow liquidity
+    pub owner: Pubkey,
+    /// Deposited collateral for the obligation, unique by deposit reserve address
+    pub deposits: Vec<ObligationCollateral>,
+    /// Borrowed liquidity for the obligation, unique by borrow reserve address
+    pub borrows: Vec<ObligationLiquidity>,
+    /// Market value of deposits
+    pub deposited_value: Decimal,
+    /// Market value of borrows
+    pub borrowed_value: Decimal,
+    /// The target ratio of borrowed value to deposited value
+    pub loan_to_value_ratio: Rate,
+    /// The loan to value ratio at which the obligation can be liquidated
+    pub liquidation_threshold: Rate,
 }
 
 impl Obligation {
-    /// Create new obligation
-    pub fn new(params: NewObligationParams) -> Self {
-        let NewObligationParams {
-            collateral_reserve,
-            borrow_reserve,
-            token_mint,
-            cumulative_borrow_rate_wads,
-        } = params;
-
-        Self {
-            version: PROGRAM_VERSION,
-            deposited_collateral_tokens: 0,
-            collateral_reserve,
-            cumulative_borrow_rate_wads,
-            borrowed_liquidity_wads: Decimal::zero(),
-            borrow_reserve,
-            token_mint,
-        }
+    /// Create a new obligation
+    pub fn new(params: InitObligationParams) -> Self {
+        let mut obligation = Self::default();
+        Self::init(&mut obligation, params);
+        obligation
     }
 
-    /// Maximum amount of loan that can be closed out by a liquidator due
-    /// to the remaining balance being too small to be liquidated normally.
-    pub fn max_closeable_amount(&self) -> Result<u64, ProgramError> {
-        if self.borrowed_liquidity_wads < Decimal::from(CLOSEABLE_AMOUNT) {
-            self.borrowed_liquidity_wads.try_ceil_u64()
-        } else {
-            Ok(0)
-        }
+    /// Initialize an obligation
+    pub fn init(&mut self, params: InitObligationParams) {
+        self.version = PROGRAM_VERSION;
+        self.last_update = LastUpdate::new(params.current_slot);
+        self.lending_market = params.lending_market;
+        self.owner = params.owner;
+        self.deposits = params.deposits;
+        self.borrows = params.borrows;
```

Comment:
nit: is there a use-case for initializing deposits and borrows?

Reply:
I do it in the tests a bunch. Does that count?

Reply:
Sure that's fair. I noticed that this always creates a stale obligation, but if it weren't for that I would be worried about this creating an obligation with an invalid state due to the values not being initialized too